### PR TITLE
Extend ILU support to AMD GPUs

### PR DESCRIPTION
**Pull request by victorapm**
Extended from #867

This PR adds HIP support to `hypre_ILU` (setup and solve phases):

1. Algorithm type 0 (BJ-ILU0)
2. Algorithm type 10 (GMRES-ILU0)
3. Iterative triangular solves for backward and forward substitutions.

TODOs for following PRs:
- ILU code refactoring
- Add regression tests

Regression tests:
- [x] tux
- [x] lassen
- [x] tioga
- [x] sunspot

### CODE REVIEW

**Comment by PaulMullowney** [comment]
With #871 merged, hopefully this can get in soon? :)

**Comment by victorapm** [comment]
Yes, Paul! I'm working on the conflicts

**Review by PaulMullowney — APPROVED** [review]
My tests all run to completion on Frontier. The performance is close to what I remember it being.

**Comment by victorapm** [comment]
Thanks for checking, Paul! I wanted to note that RCM is active by default again in this branch

**Comment by PaulMullowney** [comment]
There is something off in the performance of the "direct" solves of the ILU algorithm (i.e. when rocsparse is used for triangular solve) . I need to investigate.

**Comment by PaulMullowney** [comment]
@victorapm There is something wrong with these checks:
https://github.com/hypre-space/hypre/blob/ilu-amd/src/seq_mv/csr_matop_device.c#L2717
https://github.com/hypre-space/hypre/blob/ilu-amd/src/seq_mv/csr_matop_device.c#L2775

My solver always return false on this check and redoes the analysis at every invocation of the rocsparse triangular solve. I don't see what the issue is though.

**Comment by PaulMullowney** [comment]
I have a fix. Can I push to this branch? You can back it out/change as you see fit. My changes will need to be extended to other devices.

**Comment by victorapm** [comment]
Great!! Thanks! Yes, you should be able to push directly to this branch

**Comment by victorapm** [comment]
@rfalgout @liruipeng @oseikuffuor1 The new changes from Paul were incorporated successfully and this is ready for review again. Thanks!

File: src/seq_mv/csr_matop_device.c [hunk 1]
```diff
@@ -2044,36 +2270,43 @@ hypre_CSRMatrixSortRowOutOfPlace(hypre_CSRMatrix *A)
...
+#endif /* defined(HYPRE_USING_CUSPARSE) || defined(HYPRE_USING_ROCSPARSE) */
 
 #if defined(HYPRE_USING_CUSPARSE)
-/* @brief This functions sorts values and column indices in each row in ascending order INPLACE
- * @param[in] n Number of rows
- * @param[in] m Number of columns
- * @param[in] nnzA Number of nonzeroes
- * @param[in] *d_ia (Unsorted) Row indices
- * @param[in,out] *d_ja_sorted On Start: Unsorted column indices. On return: Sorted column indices
- * @param[in,out] *d_a_sorted On Start: Unsorted values. On Return: Sorted values corresponding with column indices
```
Comment (oseikuffuor1) [inline src/seq_mv/csr_matop_device.c:2127]: I'm curious why this was changed? I believe "@param[]" is standard syntax for documentation generating tools like oxygen?

File: src/parcsr_ls/par_ilu_solve.c [hunk 2]
```diff
@@ -1164,577 +1201,32 @@ hypre_ILUSolveLURAS(hypre_ParCSRMatrix *A,
...
-#if defined(HYPRE_USING_CUDA) && defined(HYPRE_USING_CUSPARSE)
-
-/* Permutation function (for GPU version, can just call thrust)
- * option 00: perm integer array
- * option 01: rperm integer array
- * option 10: perm real array
- * option 11: rperm real array
- * */
-HYPRE_Int
-hypre_ILUSeqVectorPerm(void      *vectori,
```
Comment (oseikuffuor1) [inline src/parcsr_ls/par_ilu_solve.c:1184]: Has this been completely removed or renamed?

**Review by oseikuffuor1 — APPROVED** [review]
Thanks Victor, this looks good. I added some minor comments for your review.

File: src/seq_mv/csr_matop_device.c (hunk 1 above)
Comment (victorapm) [reply src/seq_mv/csr_matop_device.c:2127]: Yes, right. That's for doxygen. I can revert this change. I confess I'm not paying attention to using doxygen keywords throughout the code. It's used in very few places in hypre. @rfalgout should we consider using doxygen keywords when writing new functions?

File: src/parcsr_ls/par_ilu_solve.c (hunk 2 above)
Comment (victorapm) [reply src/parcsr_ls/par_ilu_solve.c:1184]: I removed it since it's not being used anywhere in hypre

**Comment by PaulMullowney** [comment]
@liruipeng Any insight as to why
https://github.com/hypre-space/hypre/pull/877/commits/c9cb1d645fabc510fbd3f885173237df5d62b01a
is necessary? 

Prior to this commit, every time the solver entered the triangular solve algorithm, hypre_CsrsvDataInfoL (and U) return a null pointer and so the solve analysis was repeated every time, at least on AMD machines.

**Comment by victorapm** [comment]
Paul, probably I messed up something while resolving the conflicts, so that's why your fix with c9cb1d645fabc510fbd3f885173237df5d62b01a was necessary

**Comment by PaulMullowney** [comment]
> Paul, probably I messed up something while resolving the conflicts, so that's why your fix with [c9cb1d6](https://github.com/hypre-space/hypre/commit/c9cb1d645fabc510fbd3f885173237df5d62b01a) was necessary

Maybe. I think it's worth testing in master to see if the solve analysis is repeated at each invocation of the triangular solve. You can do this without ILU.  Using a gauss-seidel relaxation (3) should answer this.

**Comment by victorapm** [comment]
Yes, sounds a good idea!

File: src/seq_mv/csr_matop_device.c (hunk 1 above)
Comment (rfalgout) [reply src/seq_mv/csr_matop_device.c:2127]: We don't have a rule right now regarding parameter documentation.  We should discuss at our next meeting.  I'll put it on the agenda. :)

Comment (victorapm) [reply src/seq_mv/csr_matop_device.c:2127]: Thanks, Rob!

Comment (rfalgout) [reply src/seq_mv/csr_matop_device.c:2127]: Note that this is a non-user function, so the doxygen stuff isn't needed here.  Still something to discuss though.

**Review by rfalgout — APPROVED** [review]
This looks good to me, assuming tests are all clean.  Thanks!

File: src/parcsr_ls/par_ilu_solve.c (hunk 2 above)
Comment (liruipeng) [reply src/parcsr_ls/par_ilu_solve.c:1184]: Is `hypre_ILUSolveCusparseLU` removed?

**Comment by liruipeng** [comment]
> @liruipeng Any insight as to why [c9cb1d6](https://github.com/hypre-space/hypre/commit/c9cb1d645fabc510fbd3f885173237df5d62b01a) is necessary?
> 
> Prior to this commit, every time the solver entered the triangular solve algorithm, hypre_CsrsvDataInfoL (and U) return a null pointer and so the solve analysis was repeated every time, at least on AMD machines.

@victorapm @PaulMullowney 

I am not sure. I haven't followed all the new changes yet. What were the problems?
The original one (I put) was to check `if (!hypre_CsrsvDataInfoL(csrsv_data))`. If not, it means the csr matrix has not been analyzed for triangular solve, so we do the analysis and save in csr. I saw we have `hypre_CsrsvDataInfoLAnalyzed` now, which is fine too.

File: src/parcsr_ls/par_ilu_solve.c (hunk 2 above)
Comment (victorapm) [reply src/parcsr_ls/par_ilu_solve.c:1184]: It's been renamed to `hypre_ILUSolveLUDevice` and moved to `par_ilu_solve_device.c`

**Comment by PaulMullowney** [comment]
> > @liruipeng Any insight as to why [c9cb1d6](https://github.com/hypre-space/hypre/commit/c9cb1d645fabc510fbd3f885173237df5d62b01a) is necessary?
> > Prior to this commit, every time the solver entered the triangular solve algorithm, hypre_CsrsvDataInfoL (and U) return a null pointer and so the solve analysis was repeated every time, at least on AMD machines.
> 
> @victorapm @PaulMullowney
> 
> I am not sure. I haven't followed all the new changes yet. What were the problems? The original one (I put) was to check `if (!hypre_CsrsvDataInfoL(csrsv_data))`. If not, it means the csr matrix has not been analyzed for triangular solve, so we do the analysis and save in csr. I saw we have `hypre_CsrsvDataInfoLAnalyzed` now, which is fine too.

I think the key distinction is the the info data structs (rocsparse_mat_info) are created when the CsrsvData data structure is created. Trying to do it "on demand" within !hypre_CsrsvDataInfoL(csrsv_data) conditionals didn't enable these to persist.

My question is why and whether or not this has always been a problem.

**Comment by liruipeng** [comment]
> I think the key distinction is the the info data structs (rocsparse_mat_info) are created when the CsrsvData data structure is created. Trying to do it "on demand" within !hypre_CsrsvDataInfoL(csrsv_data) conditionals didn't enable these to persist.

Why do they not persist? They both call `HYPRE_ROCSPARSE_CALL(rocsparse_create_mat_info(&hypre_CsrsvDataInfoL(csrsv_data)) );` and save `mat_info` in the same place. I guess I am still missing something.

**Comment by victorapm** [comment]
FYI, lassen tests has some issues and I'll fix

**Comment by PaulMullowney** [comment]
> > I think the key distinction is the the info data structs (rocsparse_mat_info) are created when the CsrsvData data structure is created. Trying to do it "on demand" within !hypre_CsrsvDataInfoL(csrsv_data) conditionals didn't enable these to persist.
> 
> Why do they not persist? They both call `HYPRE_ROCSPARSE_CALL(rocsparse_create_mat_info(&hypre_CsrsvDataInfoL(csrsv_data)) );` and save `mat_info` in the same place. I guess I am still missing something.

**Agreed. For a single device run using a GS (relax_type=3) algorithm, master shows:**


A=0x2046480, csrsv_data=0x25752e0, infoL=(nil), infoU=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2494
A=0x2046480, csrsv_data=0x25752e0, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2527
A=0xd4a170, csrsv_data=0x20f2080, infoL=(nil), infoU=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2494
A=0xd4a170, csrsv_data=0x20f2080, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2527
A=0xa38390, csrsv_data=0x20e9fe0, infoL=(nil), infoU=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2494
A=0xa38390, csrsv_data=0x20e9fe0, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2527
    1    9.299483e+01    0.053174   5.317365e-02
    2    2.255780e+01    0.242570   1.289836e-02
    3    3.842656e+00    0.170347   2.197198e-03
    4    7.845689e-01    0.204174   4.486098e-04
    5    1.701871e-01    0.216918   9.731153e-05
    6    3.864283e-02    0.227061   2.209564e-05
    7    8.583455e-03    0.222123   4.907946e-06
    8    1.952554e-03    0.227479   1.116454e-06
    9    4.204593e-04    0.215338   2.404150e-07
   10    9.019544e-05    0.214516   5.157298e-08
   
   **When checking on (!hypre_CsrsvDataInfoL(csrsv_data)), ilu-amd shows:**
   
A=0x2043d00, csrsv_data=0x2592330, infoL=(nil), infoU=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2695
A=0x2043d00, csrsv_data=0x2592330, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x2043d00, csrsv_data=0x2592330, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x20f85a0, csrsv_data=0x25ecaa0, infoL=(nil), infoU=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2695
A=0x20f85a0, csrsv_data=0x25ecaa0, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x20f85a0, csrsv_data=0x25ecaa0, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x25f1df0, csrsv_data=0x13e3240, infoL=(nil), infoU=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2695
A=0x25f1df0, csrsv_data=0x13e3240, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x25f1df0, csrsv_data=0x13e3240, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x25f1df0, csrsv_data=0x13e3240, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x25f1df0, csrsv_data=0x13e3240, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x20f85a0, csrsv_data=0x25ecaa0, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x20f85a0, csrsv_data=0x25ecaa0, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x2043d00, csrsv_data=0x2592330, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x2043d00, csrsv_data=0x2592330, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
    1    9.299483e+01    0.053174   5.317365e-02
A=0x2043d00, csrsv_data=0x2592330, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x2043d00, csrsv_data=0x2592330, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x20f85a0, csrsv_data=0x25ecaa0, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x20f85a0, csrsv_data=0x25ecaa0, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x25f1df0, csrsv_data=0x13e3240, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x25f1df0, csrsv_data=0x13e3240, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x25f1df0, csrsv_data=0x13e3240, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x25f1df0, csrsv_data=0x13e3240, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x20f85a0, csrsv_data=0x25ecaa0, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x20f85a0, csrsv_data=0x25ecaa0, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x2043d00, csrsv_data=0x2592330, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
A=0x2043d00, csrsv_data=0x2592330, infoL=(nil) : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2733
    2    2.255780e+01    0.242570   1.289836e-02

**Comment by PaulMullowney** [comment]
Even my fixes don't seem to be working on Frontier. I made the following changes:
````
diff --git a/src/seq_mv/csr_matop_device.c b/src/seq_mv/csr_matop_device.c
index 6b08ed489..2aab1768e 100644
--- a/src/seq_mv/csr_matop_device.c
+++ b/src/seq_mv/csr_matop_device.c
@@ -2727,8 +2727,14 @@ hypre_CSRMatrixTriLowerUpperSolveRocsparse(char              uplo,
       HYPRE_ROCSPARSE_CALL( rocsparse_set_mat_fill_mode(descr, rocsparse_fill_mode_lower) );

       /* TODO (VPM): move the following block to hypre_CSRMatrixTriSetupRocsparse */
+               printf("\nA=%p, csrsv_data=%p, infoL=%p, analyzed=%d : %s %s %d\n",A,csrsv_data,
+                                hypre_CsrsvDataInfoL(csrsv_data),hypre_CsrsvDataAnalyzedL(csrsv_data),
+                                __FILE__,__FUNCTION__,__LINE__);
       if (!hypre_CsrsvDataAnalyzedL(csrsv_data))
       {
+                       printf("A=%p, csrsv_data=%p, infoL=%p, analyzed=%d : %s %s %d\n",A,csrsv_data,
+                                        hypre_CsrsvDataInfoL(csrsv_data),hypre_CsrsvDataAnalyzedL(csrsv_data),
+                                        __FILE__,__FUNCTION__,__LINE__);
          HYPRE_ROCSPARSE_CALL( hypre_rocsparse_csrsv_buffer_size(handle,
                                                                  rocsparse_operation_none,
                                                                  num_rows, num_nonzeros, descr,
````
THen I compiled with 5.4.3 and 5.4.0. In both cases, I see:

A=0xf76e60, csrsv_data=0x27db180, infoL=0x24b5170, **analyzed=1** : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2732
A=0xf76e60, csrsv_data=0x27db180, infoL=0x24b5170, **analyzed=0** : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2737

I need help with this.

**Comment by victorapm** [comment]
Hi Paul, do you have a reproducer that uses the ij driver?

**Comment by PaulMullowney** [comment]
./ij -rlx 3 -solver 3

A=0x1b58ab0, csrsv_data=0x25936b0, infoL=0x25940c0, analyzed=1 : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2732
A=0x1b58ab0, csrsv_data=0x25936b0, infoL=0x25940c0, analyzed=0 : csr_matop_device.c hypre_CSRMatrixTriLowerUpperSolveRocsparse 2737

**Comment by victorapm** [comment]
I'll take a look now

**Comment by PaulMullowney** [comment]
@victorapm Can you reproduce?

**Comment by victorapm** [comment]
Yes, I can reproduce it. Trying to figure out what's happening now...

**Comment by PaulMullowney** [comment]
Ok. Good. I thought I was losing my mind on this

**Comment by victorapm** [comment]
Yes, that's a very weird behavior going on in the analysis phase...

**Comment by liruipeng** [comment]
> Yes, that's a very weird behavior going on in the analysis phase...

Is the sorting the reason?

**Comment by victorapm** [comment]
Don't know yet. Still trying to figure out

**Comment by victorapm** [comment]
@PaulMullowney @liruipeng It is fixed now, i.e., the analysis phase is executed only once. Can you confirm in your end, Paul? It seems that one of the pre-processor macros was expanding in a weird way and the following `if` condition was skip. I know this is weird, but it's fixed from my side :)

**Comment by liruipeng** [comment]
> @PaulMullowney @liruipeng It is fixed now, i.e., the analysis phase is executed only once. Can you confirm in your end, Paul? It seems that one of the pre-processor macros was expanding in a weird way and the following `if` condition was skip. I know this is weird, but it's fixed from my side :)

Hi @victorapm . Thank you for the fix. I looked at the commit but I didn’t really understand. sorry

**Comment by PaulMullowney** [comment]
> > @PaulMullowney @liruipeng It is fixed now, i.e., the analysis phase is executed only once. Can you confirm in your end, Paul? It seems that one of the pre-processor macros was expanding in a weird way and the following `if` condition was skip. I know this is weird, but it's fixed from my side :)
> 
> Hi @victorapm . Thank you for the fix. I looked at the commit but I didn’t really understand. sorry

I don't understand it either but it seems to fix the problem :(

**Comment by victorapm** [comment]
Hi all, here's a better explanation of the fix:

In the old code, `buffer` could be used uninitialized, see the compiler warning:
```
csr_matop_device.c:2742:11: warning: variable 'buffer' is used uninitialized whenever 'if' condition is false [-Wsometimes-uninitialized]
      if (!hypre_CsrsvDataAnalyzedL(csrsv_data))
          ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
csr_matop_device.c:2793:57: note: uninitialized use occurs here
                                                        buffer) );
                                                        ^~~~~~
```

This was line 2742 for your reference:
```
      if (!hypre_CsrsvDataAnalyzedL(csrsv_data))
```

So, to avoid the problem of using `buffer` uninitialized later on, the compiler decided to be "smart" and compiled the above `if` condition as `if (1)`, i.e., always true. Note that, in this case `buffer` is initialized in the following lines of code just before it's used.

The fix boiled down to simply initializing `buffer` before line 2742. Even `buffer = NULL` would be sufficient to make the compiler not do his smart trick.

Clearly, we were not being informed properly of this `if (1)` trick by the compiler. I found it out using the debugger (`rocgdb`).

Hope this helps!
Victor

**Comment by liruipeng** [comment]
> Hi all, here's a better explanation of the fix:
> 
> In the old code, `buffer` could be used uninitialized, see the compiler warning:
> 
> ```
> csr_matop_device.c:2742:11: warning: variable 'buffer' is used uninitialized whenever 'if' condition is false [-Wsometimes-uninitialized]
>       if (!hypre_CsrsvDataAnalyzedL(csrsv_data))
>           ^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
> csr_matop_device.c:2793:57: note: uninitialized use occurs here
>                                                         buffer) );
>                                                         ^~~~~~
> ```
> 
> This was line 2742 for your reference:
> 
> ```
>       if (!hypre_CsrsvDataAnalyzedL(csrsv_data))
> ```
> 
> So, to avoid the problem of using `buffer` uninitialized later on, the compiler decided to be "smart" and compiled the above `if` condition as `if (1)`, i.e., always true. Note that, in this case `buffer` is initialized in the following lines of code just before it's used.
> 
> The fix boiled down to simply initializing `buffer` before line 2742. Even `buffer = NULL` would be sufficient to make the compiler not do his smart trick.
> 
> Clearly, we were not being informed properly of this `if (1)` trick by the compiler. I found it out using the debugger (`rocgdb`).
> 
> Hope this helps! Victor

Compilers shouldn't be that smarter than humans...

**Review by rfalgout — APPROVED** [review]
This all looks good to me, as long as the runs are all clean.  This was a lot of work!  Thanks!

**Review by rfalgout — APPROVED** [review]

**Comment by victorapm** [comment]
Thanks, Rob! TEST_ams from the cuda-12 lassen build is the only one failing, but it should be fixed after we merge #924

**Comment by PaulMullowney** [comment]
@victorapm Thank you so much for all the working in getting this PR ready for master. I know it was a slog. I think it's a great new capability that several apps will benefit from.

**Comment by victorapm** [comment]
Sure, Paul! Thank you for all your contributions and help in debugging as well! Most of the new code here comes from you :) Hopefully many apps will benefit from this!

**Comment by rfalgout** [comment]
Hi @victorapm .  Before you merge this, would you mind running astyle?  There are a few files on `master` currently that are not passing.  I figure this is easier than creating a separate PR.  Thanks!

**Comment by victorapm** [comment]
Yes, sure! I've done that on my local branch already and plan to push after we merge #924

**Review by liruipeng — APPROVED** [review]
Looks good. Thanks!

**Comment by PaulMullowney** [comment]
Hallelujah!!! :)

**Pull request merged**